applications/nrf_desktop/src/hw_interface/selector_hw.c

```diff
@@ -0,0 +1,352 @@
+{
+	struct selector *selector = CONTAINER_OF(w, struct selector, work);
+	const struct gpio_pin *sel_pins = selector->config->pins;
+
+	for (size_t i = 0; i < selector->config->pins_size; i++) {
+		if ((sel_pins[i].port == selector->read_port) &&
+		    (selector->read_pins_mask & BIT(sel_pins[i].pin))) {
+			/* Only one pin may be set at the time. */
+			if (selector->read_pins_mask & ~BIT(sel_pins[i].pin)) {
+				state = STATE_ERROR;
+			}
+			selector->state = i;
+			break;
+		}
+	}
+
+	if (state == STATE_ERROR) {
+		module_set_state(MODULE_STATE_ERROR);
+		return;
+	}
+
+	selector_event_send(selector);
+
+	int err = enable_interrupts_lock(selector);
+
+	if (err) {
+		module_set_state(MODULE_STATE_ERROR);
+	}
+}
+
+static int configure(struct selector *selector)
+{
+	const struct gpio_pin *sel_pins = selector->config->pins;
+	u32_t bitmask = 0;
+
+	for (size_t i = 0; i < ARRAY_SIZE(gpio_dev); i++) {
+		for (size_t j = 0; j < selector->config->pins_size; j++) {
+			if (sel_pins[j].port == i) {
+				bitmask |= BIT(sel_pins[j].pin);
+			}
+		}
+		gpio_init_callback(&selector->gpio_cb[i], selector_isr,
+				   bitmask);
+		int err = gpio_add_callback(gpio_dev[i], &selector->gpio_cb[i]);
+
+		if (err) {
+			LOG_ERR("Cannot add callback (err %d)", err);
+			return err;
+		}
+
+		bitmask = 0;
+	}
+
+	for (size_t i = 0; i < selector->config->pins_size; i++) {
+		int err = gpio_pin_configure(gpio_dev[sel_pins[i].port],
+					     sel_pins[i].pin,
+					     GPIO_PUD_PULL_DOWN | GPIO_DIR_IN |
+					     GPIO_INT | GPIO_INT_LEVEL |
+					     GPIO_INT_ACTIVE_HIGH);
+
+		if (err) {
+			LOG_ERR("Cannot configure pin (err %d)", err);
+			return err;
+		}
+	}
+
+	k_delayed_work_init(&selector->work, selector_work_fn);
+
+	return 0;
+}
+
+static int read_state_and_enable_interrupts(struct selector *selector)
+{
+	int err = read_state(selector);
+
+	if (err) {
+		return err;
+	}
+
+	err = enable_interrupts_lock(selector);
+
+	return err;
+}
+
+static int init(void)
+{
+	int err;
+
+	for (size_t i = 0; i < ARRAY_SIZE(port_map); i++) {
+		gpio_dev[i] = device_get_binding(port_map[i]);
+
+		if (!gpio_dev[i]) {
+			LOG_ERR("Cannot get GPIO%u dev binding", i);
+			return -ENXIO;
+		}
+	}
+
+	for (size_t i = 0; i < ARRAY_SIZE(selectors); i++) {
+		selectors[i].config = &selector_configs[i];
+		err = configure(&selectors[i]);
+		if (err) {
+			return err;
+		}
+		err = read_state_and_enable_interrupts(&selectors[i]);
+		if (err) {
+			return err;
+		}
+	}
+	state = STATE_ACTIVE;
+
+	return 0;
+}
+
+static void sleep(void)
+{
+	for (size_t i = 0; i < ARRAY_SIZE(selectors); i++) {
+		int err = disable_interrupts_nolock(&selectors[i]);
+
+		if (err) {
+			module_set_state(MODULE_STATE_ERROR);
+			return;
+		}
+
+		k_delayed_work_cancel(&selectors[i].work);
+	}
+
+	state = STATE_OFF;
+	module_set_state(MODULE_STATE_OFF);
+}
+
+static void wake_up(void)
+{
+	for (size_t i = 0; ARRAY_SIZE(selectors); i++) {
+		int err = read_state_and_enable_interrupts(&selectors[i]);
```

Comment:
I would prefer triggering work here

Reply:
you can also call a work function directly

Reply:
After current changes, work calls `read_state_and_enable_interrupts`. Isn't calling this function here more efficient and readable than triggering work (function name clearly suggests what is done)?